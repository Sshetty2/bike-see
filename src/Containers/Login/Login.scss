.overlay-div {
  position: fixed;
  background-color: rgba(0, 0, 0, 0.5);
  left: 0;
  top: 0;
  width: 100%;
  min-height: 100%;
  z-index: 10;
}

.login-pop-up {
  position: fixed;
  width: 300px;
  height: 450px;
  background-color: $background-color;
  border-radius: 10px;
  padding: 20px;
  margin-left: -150px;
  margin-top: -230px;
  left: 50%;
  top: 50%; 
  box-shadow: 2px 2px 8px black;
  z-index: 10;
}

h3 {
  margin: 10px;
  text-align: center;
  font-size: 24px;
  color: $header-text-color;
  font-weight: 800;
}