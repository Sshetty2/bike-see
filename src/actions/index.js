export const setLoading = (loading) => ({
  type: 'SET_LOADING',
  loading,
});

export const setError = (error) => ({
  type: 'SET_ERROR',
  error,
});

export const setCities = (cities) => ({
  type: 'SET_CITIES',
  cities,
});

export const setCurrentCity = (city) => ({
  type: 'SET_CURRENT_CITY',
  city,
});

export const setStations = (stations) => ({
  type: 'SET_STATIONS',
  stations,
});

export const toggleFavorite = (favorite) => ({
  type: 'TOGGLE_FAVORITE',
  favorite,
});