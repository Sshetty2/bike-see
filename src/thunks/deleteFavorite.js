import { removeFavorite, setLoading, setError } from '../actions';
import { fetchData } from '../utils/api';

export const deleteFavorite = (station_id, user_id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`https://bike-see.herokuapp.com/api/v1/users/${user_id}/favorites/${station_id}`, 'DELETE');
      dispatch(removeFavorite(station_id));
    } catch (error) {
      dispatch(setError('Error deleting favorite.'));
    }
    dispatch(setLoading(false));
  }
}