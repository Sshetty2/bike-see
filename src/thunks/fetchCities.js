import { setLoading, setError, setCities } from '../actions';
import { fetchData } from '../utils/api';

export const fetchCities = () => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const results = await fetchData('https://api.citybik.es/v2/networks', 'GET');
      dispatch(setCities(results.networks));
    } catch (error) {
      dispatch(setError('Error fetching cities.'));
    }
    dispatch(setLoading(false));
  }
}