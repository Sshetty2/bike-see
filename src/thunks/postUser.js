import { setCurrentUser, setLoading, setError } from '../actions';
import { fetchData } from '../utils/api';

export const postUser = (user) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const fetchedUser = await fetchData('https://bike-see.herokuapp.com/api/v1/users/new', 'POST', user);
      dispatch(setCurrentUser(fetchedUser));
      localStorage.setItem('bike-user', JSON.stringify(fetchedUser));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}