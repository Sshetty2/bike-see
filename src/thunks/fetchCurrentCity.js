import { setLoading, setError, setCurrentCity } from '../actions';
import { fetchStations } from '../thunks/fetchStations';
import { fetchData } from '../utils/api';

export const fetchCurrentCity = (user_id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const fetchedCity = await fetchData(`http://localhost:3001/api/v1/users/${user_id}/city`, 'GET');

      if (fetchedCity !== '') {
        dispatch(setCurrentCity(fetchedCity.city));
        dispatch(fetchStations(user_id, fetchedCity.city));
      }

    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}