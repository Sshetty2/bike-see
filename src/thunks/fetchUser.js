import { setCurrentUser, setLoading, setError } from '../actions';
import { fetchCurrentCity } from '../thunks/fetchCurrentCity';
import { fetchFavorites } from '../thunks/fetchFavorites';
import { fetchData } from '../utils/api';

export const fetchUser = (user) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const fetchedUser = await fetchData('http://localhost:3001/api/v1/users', 'POST', user);
      dispatch(setCurrentUser(fetchedUser));
      localStorage.setItem('bike-user', JSON.stringify(fetchedUser));
      dispatch(fetchFavorites(fetchedUser.id));
      dispatch(fetchCurrentCity(fetchedUser.id));
    } catch (error) {
      dispatch(setError('Error logging in user. Please sign up or try again.'));
    }
    dispatch(setLoading(false));
  }
}