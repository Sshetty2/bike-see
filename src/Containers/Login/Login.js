import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { fetchUser } from '../../thunks/fetchUser';
import { connect } from 'react-redux';

class Login extends Component {
  constructor() {
    super();
    this.state = {
      user: {
        name: '',
        email: '',
        password: '',
        verifyPassword: '',
      },
      response: ''
    }
  }

  updateState = (e) => {
    e.preventDefault();
    const { id, value } = e.target;
    this.setState({ [id]: value })
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { login, fetchUser } = this.props;
    fetchUser(this.state.user, login);
  }

  render() {
    const { login } = this.props;
    const { user } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <label>Email
          <input onChange={this.updateState} required type='text' placeholder='Enter your email' value={user.email} id='email'/>
        </label>
        <label>Password
          <input onChange={this.updateState} required type='password' placeholder='Enter your password' value={user.password} id='password'/>
        </label>
        {!login &&
          <label>Verify Password
            <input onChange={this.updateState} required type='password' placeholder='Enter your password' value={user.verifyPassword} id='verifyPassword'/>
          </label>
        }
        <button type='submit'>Submit</button>
        {login ? 
          <Link className='pop-up-link' to='/sign-up'>Sign Up Here</Link> : 
          <Link className='pop-up-link' to='/login'>Login Here</Link>
        }
        <p className='error-text hidden'>Please try again</p>
      </form>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  fetchUser: (user) => dispatch(fetchUser(user)),
});

export default connect(null, mapDispatchToProps)(Login);