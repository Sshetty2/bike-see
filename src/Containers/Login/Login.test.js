import React from 'react';
import { Login, mapStateToProps, mapDispatchToProps } from './Login';
import { shallow } from 'enzyme';
import { setError } from '../../actions';
import { fetchUser } from '../../thunks/fetchUser';
import { postUser } from '../../thunks/postUser';
jest.mock('../../thunks/fetchUser');
jest.mock('../../thunks/postUser');

describe('Login', () => {
  let wrapper;
  let fetchUserMock;
  let postUserMock;
  let setErrorMock;
  let mockUser;
  let mockError;
  let mockEvent;
  let mockHistory;

  beforeEach(() => {
    fetchUserMock = jest.fn();
    postUserMock = jest.fn();
    setErrorMock = jest.fn();
    mockError = '';
    mockHistory = { replace: jest.fn() };
    mockUser = { id: '1', name: 'Bobby', email: 'b@b', password: 'b'};
    wrapper = shallow(
      <Login
        error={mockError}
        user={mockUser}
        fetchUser={fetchUserMock}
        setError={setErrorMock}
        postUser={postUserMock}
        login={true}
        history={mockHistory}
      />
    )
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual({
      name: '',
      email: '',
      password: '',
      verifyPassword: ''
    });
  });

  describe('clearError', () => {
    it('should call setError', () => {
      wrapper.instance().clearError();
      expect(wrapper.instance().props.setError).toHaveBeenCalled();
    });
  });

  describe('updateState', () => {
    it('should update state with new values', () => {
      mockEvent = {
        preventDefault: jest.fn(),
        target: { id: 'email', value: 'C@C' }
      };
      const expected = 'c@c';
      wrapper.instance().updateState(mockEvent);
      expect(wrapper.state('email')).toEqual(expected);
    });

    it('should update more state with new values', () => {
      mockEvent = {
        preventDefault: jest.fn(),
        target: { id: 'name', value: 'Karin' }
      };
      const expected = 'Karin';
      wrapper.instance().updateState(mockEvent);
      expect(wrapper.state('name')).toEqual(expected);
    });
  });

  describe('handleSubmit', () => {
    it('should call loginUser with the correct params', async () => {
      wrapper.instance().loginUser = jest.fn();
      await wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.instance().loginUser).toHaveBeenCalled();
    });

    it('should call signUpUser with the correct params', async () => {
      wrapper = shallow(
        <Login
          user={mockUser}
          fetchUser={fetchUserMock}
          postUser={postUserMock}
          history={mockHistory}
          login={false}
        />
      );
      wrapper.instance().signUpUser = jest.fn();
      await wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.instance().signUpUser).toHaveBeenCalled();
    });

    it('should call updatePath', async () => {
      wrapper.instance().clearError = jest.fn();
      wrapper.instance().updatePath = jest.fn();
      await wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.instance().updatePath).toHaveBeenCalled();
    });
  });

  describe('signUpUser', () => {
    it('should call postUser with the correct params', async () => {
      await wrapper.instance().signUpUser(mockEvent);
      expect(wrapper.instance().props.postUser).toHaveBeenCalledWith({ name: '', email: '', password: '' });
    });

    it('should call setError if passwords dont match', async () => {
      wrapper.setState({ password: 'bleep', verifyPassword: 'bloop'});
      await wrapper.instance().signUpUser(mockEvent);
      expect(wrapper.instance().props.setError).toHaveBeenCalled();
    });
  });

  describe('loginUser', () => {
    it('should call fetchUser with the correct params', async () => {
      await wrapper.instance().loginUser(mockEvent);
      expect(wrapper.instance().props.fetchUser).toHaveBeenCalledWith({ email: '', password: '' });
    });
  });

  describe('updatePath', () => {
    it('should call history.replace', () => {
      wrapper.instance().updatePath();
      expect(wrapper.instance().props.history.replace).toHaveBeenCalled();
    });

    it('should not call history.replace if there is an error', () => {
      mockError = 'There is an error';
      wrapper = shallow(
        <Login
          error={mockError}
          history={mockHistory}
        />
      );
      wrapper.instance().updatePath();
      expect(wrapper.instance().props.history.replace).not.toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object of props', () => {
      const mockState = {
        error: mockError,
        user: mockUser,
        otherState: false,
      };
      const expected = {
        error: mockError,
        user: mockUser,
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when fetchUser is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchUser();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchUser();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when setError is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setError();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setError();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when postUser is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = postUser();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.postUser();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});