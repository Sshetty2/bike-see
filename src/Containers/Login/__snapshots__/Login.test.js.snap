// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the correct snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    Email
    <input
      id="email"
      onChange={[Function]}
      placeholder="Enter your email"
      required={true}
      type="email"
    />
  </label>
  <label>
    Password
    <input
      id="password"
      onChange={[Function]}
      placeholder="Enter your password"
      required={true}
      type="password"
    />
  </label>
  <button
    type="submit"
  >
    Submit
  </button>
  <Link
    className="pop-up-link"
    onClick={[Function]}
    replace={false}
    to="/sign-up"
  >
    Sign Up Here
  </Link>
  <p
    className="error-text"
  />
</form>
`;
